# TIGER cASE analysis repository
Combined ASE expression scripts used in the Alonso _et al._ TIGER paper.

The results generated by the scripts available in this repository, plus 
all other results generated in this study, can be explored in the
[TIGER Data Portal](http://tiger.bsc.es), alongside with other
relevant genomic, epigenetic and transcriptomic data.

## Preface
This repository contains the C++ source code for four scripts that
were used in different analyses in the study.

Each script can be compiled to a binary executable file by using

```g++ -Wall -o step01_binary step_01.cpp ../utils/allelic_imba_common.cpp```

Each binary expects a config.ini file residing in the same directory,
containing the necessary arguments for the executions. Example config.ini
files are already included in this repository, with a brief description of 
the type of the expected arguments, and of their particular formatting.
Prior to executing the binary, please _make sure that all the necessary 
arguments in config.ini are correct_ and that they refer to the expected 
data files. Otherwise, the execution will terminate unexpectedly.

The code is provided as is, for informational purposes and without any 
warranties. It has only been tested in a limited number of environments.

## 01 Artificial Reads
This script reads a list of SNPs and a fasta genome, and generates
all possible reads overlapping the SNPs, containing the same number of 
reads containing the reference and alternate alleles. It also outputs
a histogram of the SNP density in each base, and a list of regions
discarded due to excessive density of polymorphisms.

A set of artificial RNAseq reads generated with this code was used to 
test for alignment allelic biases in all possible reporter variants. Given 
that the same number of reads are generated containing the reference and 
alternate alleles, any deviation form the expected 50% allelic ratio can be
attributed to problems with the alignment strategy.

## 02 Filter and Unify Alignments
This script reads, for all samples listed as input argument, all allelic 
imbalance quantification.txt files obtained as explained in the Online Methods
secton of the paper. It also requires a .vcf genotype file, with the genotypes 
of all samples of interest in columns in any order. This script then
filters all input data and creates a unified output file with all the 
allelic imbalance information, along with other files on mean allelic biases,
or Benjamini Hochberg significance thresholds.

The resulting outputs are intermediary files necessary for the downstream 
analyses.

## 03 Permuted Reporter Imbalance
This script reads an Allelic_Imba_trimmed.txt file, outputed in step 02,
calculates the empiric Zscore distribution, then randomises
the Het read counts N times, and creates an expected null Zscore 
distribution. The counts are randomised within 5 gene expression bins, 
one for SNPs with a median read coverage of zero, and four more in increasing 
quartiles for the remainder of the data.

The Z-score distribution obtained from these distributions is used 
in the downstream analyses as the null distribution, ie the expected
distribution of allelic imbalance Z-scores that would be obtained if 
all allelic biases were the product of biological stochasticity.




