# TIGER_cASE
Combined ASE expression scripts used in the Alonso et al TIGER paper.

The results generated by the scripts available in this repository, plus 
all other results generated in this study, can be explored in the
[TIGER Data Portal](http://tiger.bsc.es), alongside with other
relevant genomic, epigenetic and transcriptomic data.

## Preface
This repository contains the C++ source code for four scripts that
were used in different analyses in the study.

Each script can be compiled to a binary executable file by using
```g++ -Wall -fexceptions -O2 -o step01_binary step_01.cpp ../utils/utils.cpp```

The code is provided as is, for informational purposes and without any 
warranties. It has only been tested in a limited number of environments.

## 01 Artificial Reads
This script reads a list of SNPs and a fasta genome, and generates
all possible reads overlapping the SNPs, containing the same number of 
reads containing the reference and alternate alleles. It also outputs
a histogram of the SNP density in each base, and a list of regions
discarded due to excessive density of polymorphisms.

A set of artificial RNAseq reads generated with this code was used to 
test for alignment allelic biases in all possible reporter variants. Given 
that the same number of reads are generated containing the reference and 
alternate alleles, any deviation form the expected 50% allelic ratio can be
attributed to problems with the alignment strategy.

